### ISSUE TITLE
节点智能选择建议：Proposal for Advanced Automatic Node Selection in sing-box

### ISSUE THREAD

**Issue by dbsxdbsx** [SagerNet/sing-box#2061]
I think sing-box currently lacks a robust, built-in mechanism for intelligent node selection and failover that can work autonomously without external APIs while respecting user-defined priorities.

Current methods and their limitations(correct me if I am wrong):

1. "selector" type:
   - Requires Clash API for control
```json
{
  "type": "selector",
  "tag": "select_outbound",
  "outbounds": ["nodeA", "nodeB", "nodeC"],
  "default": "nodeA",
  "interrupt_exist_connections": false
}
```

2. "urltest" type:
   - Selects based solely on latency to a specific URL(or the default one if field `url` is not set)
   - Doesn't guarantee accessibility to actual target URLs
   - May not reflect real-world performance for diverse traffic

```json
{
  "type": "urltest",
  "tag": "auto_select",
  "outbounds": ["nodeA", "nodeB", "nodeC"],
  "url": "https://www.gstatic.com/generate_204",
  "interval": "10m",
  "tolerance": 50
}
```

#### Proposed Solution
Implement a new outbound type "smart_failover" with the following features:

1. Works natively within sing-box
2. Allows defining priority order for nodes
3. Attempts connections in the order specified in step 2 until a working node is found
4. Give option to control whether need to periodically checks preferred nodes to switch back when available

Suggested configuration:

```json
{
  "type": "smart_failover",
  "tag": "smart_outbound",
  "outbounds": ["nodeA", "nodeB", "nodeC"],
  "failover_timeout": "5s",
  "check_interval": "5m",
  "health_check": {
    "url": "https://www.gstatic.com/generate_204",
    "timeout": "2s"  // if timouted, it means the node is not even pass the basic health check
  }
}
```

Here, for all candidate nodes, the proposed solution would first perform a health check, eliminating those nodes that fail to connect after the specified `timeout` period. Then, for the remaining nodes, it would check them in the original order specified in the `outbounds` field until a working node is found. In this context, a working node is defined as a node that can successfully connect to the target URL within the time defined in `failover_timeout`. Additionally, if the `check_interval` is provided, the solution would periodically check the preferred nodes to switch back when they become available. This behavior is similar to the `interrupt_exist_connections` feature in the `"type": "selector"` configuration.

So in a word, I hope there could be a kind of outbound config, that could make the picked proxy server node more diversified for each actual net stream .

**Comment by shakibamoshiri** [comment]
I think there is no need for defining a new outbound.  
if you have tested Xray , this is done by default. When there are two outbounds, Xray checks if can connect to them; and if one outbound can be available, it routes requests to the available outbound

As I tested singbox 1.9 and adding **URLTest** to the outbound list, it can detect that an outbound is not available while the second one does; but the failover or  availability selection does not happen


```bash
INFO[0000] outbound/hysteria2[proxy2]: outbound connection to www.gstatic.com:443
DEBUG[0000] ...
DEBUG[0000] ...
INFO[0000] outbound/hysteria2[proxy]: outbound connection to www.gstatic.com:443
DEBUG[0000] outbound/urltest[auto]: outbound proxy unavailable: INTERNAL_ERROR (local): read ...
DEBUG[0000] outbound/direct[direct]: outbound packet connection to 1.1.1.2:53
DEBUG[0000] outbound/direct[direct]: outbound packet connection to 1.1.1.2:53
DEBUG[0000] ...
DEBUG[0001] outbound/urltest[auto]: outbound proxy2 available: 408ms
```

so sing-box already knows which outbound is available and can route to the healthy one

**Comment by dbsxdbsx** [comment]
> I think there is no need for defining a new outbound. if you have tested Xray , this is done by default. When there are two outbounds, Xray checks if can connect to them; and if one outbound can be available, it routes requests to the available outbound
> 
> As I tested singbox 1.9 and adding **URLTest** to the outbound list, it can detect that an outbound is not available while the second one does; but the failover or availability selection does not happen
> 
> INFO[0000] outbound/hysteria2[proxy2]: outbound connection to www.gstatic.com:443
> DEBUG[0000] ...
> DEBUG[0000] ...
> INFO[0000] outbound/hysteria2[proxy]: outbound connection to www.gstatic.com:443
> DEBUG[0000] outbound/urltest[auto]: outbound proxy unavailable: INTERNAL_ERROR (local): read ...
> DEBUG[0000] outbound/direct[direct]: outbound packet connection to 1.1.1.2:53
> DEBUG[0000] outbound/direct[direct]: outbound packet connection to 1.1.1.2:53
> DEBUG[0000] ...
> DEBUG[0001] outbound/urltest[auto]: outbound proxy2 available: 408ms
> so sing-box already knows which outbound is available and can route to the healthy one

I think you missed my point.  The key is not if singbox could connect to a specific outbound node with lowest legacy, nor if the singbox `urltest` node could fetch the most feasible and fast outbound node to the URL set in field `"url":` of the `urltest` outbound , the key is how to make the `urltest` node automatically adaptive to different real URL, for example: 

if I am surfing youtube, then the test with the default `"https://www.gstatic.com/generate_204"` URL is meaningless, but `www.youtube.com`  is meaningful.  

I've tried to simplify this by adding  many URL test nodes by my customized app:



But still this work is a little tedious, since once there is a new website needs to proxy, I need to add it.

Plz directly tell me if I missed something.

**Comment by shakibamoshiri** [comment]
@dbsxdbsx 
me neither am sure if understood your fully and correctly 

your comment is what got my attention 

> I think sing-box currently lacks a robust, built-in mechanism for intelligent node selection and failover 

I was reading the docs to find a way of handling failover . And by failover I mean setting up a kind of (HA = High Availability) since I did it with Xray-Core .

In Xray-Core we can define two **outbounds** which both are connected an functional, but if one outbound disconnected  (e.g. server down-time)  then the other outbound is selected by Xray-Core handing all user requested. (a kind of load balancing)

Connecting to an endpoint with the lowest latency is a **nice to have feather** and I did not talk about this. My main point is having a backup outbound in case of failure of the active (primary) one

Also **HaPorxy** has this feature which a user can define two endpoints, one as an active endpoint and the other as a backup. If the active one health check fails , HaProxy sends all the requests to the backup endpoint

### why VPN/Porxy servers need this feature?
it is simple, the blockage. We live in countries with high censorship and suppression   . We wake up morning and the server IP is fully blocked.  
A normal user/admin would go and buy another server, but as a DevOps/SRE I see it differently (HA is the solution)

Thank you

**Comment by dbsxdbsx** [comment]
@shakibamoshiri , I still think we are talking 2 topics not the same--- you focus on the backon with the metric of whether the chosen outbound server is connectable, by which I think the current singbox urltest node can meet what you need by just put more than 1 outbound like this `"outbounds": ["nodeA", "nodeB", "nodeC"],` field, example:
```json
{
  "type": "urltest",
  "tag": "auto_select",
  "outbounds": ["nodeA", "nodeB", "nodeC"],
  "url": "https://www.gstatic.com/generate_204",
  "interval": "10m",
  "tolerance": 50
}
```

but I  focus on the final  performance to the target URL, by which the `  "url": "https://www.gstatic.com/generate_204",` is the key field  in the current urltest .

**Comment by shakibamoshiri** [comment]
@dbsxdbsx thank you considering replying to me , I am glad to read up again and got your point (hopefully)

First I did a retest and realized sing-box has that feature and can HA between two outbounds. 👍 

Second I think the whole point of the "url" as I tested  is to measure node latency and pick one with the lowest one. And by saying  **node** I mean our exit node (= VPN/Porxy server). Now that I am testing it, it does not make make sense to me  for example set "url" : "youtube.com" if I am considering HA and failover handling 
Even the "url" should be my own so I can be sure about its latency time 


but your point is that if a user a more connecting/checking i.e youtube.com,  then "url" is better to adopt auto-magically to "youtube.com", a kind High Frequency Visit Site be set for the "url" (?)
And if this is your point, and the above feature does not solve the issue (failover) for you, then what issues are you trying to solve by this proposal ?
